O3 tooltip (#1429)
Financial-Times/origami: components/o3-tooltip/README.md

# o3-tooltip[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](#licence)

o3-tooltip is a component usually used for annotating or highlighting bits of user interface.

- [o3-tooltip](#o3-tooltip)
	- [Markup](#markup)
		- [Branding](#branding)
	- [JavaScript](#javascript)
	- [css](#css)
	- [JSX](#jsx)
	- [Migration Guide](#migration-guide)
	- [Contact](#contact)
	- [Licence](#licence)

## Markup

> `o3-tooltip` currently supports two variants, which is determined by `render-on-open` attribute on custom element. If attribute is enable the tooltip will be rendered once the DOM is loaded and will have a close button. Close button remove the tooltip from the DOM and it can re-triggered only reload.

Reviewer: `render-on-open` is a behaviour driven API. This reminds me of the `o-tooltip` API where it was possible to configure lots of behaviours. However, that lead to lots of misuse.

It's difficult to talk communicate the "tooltip without render-on-open enabled" vs. "toggletip", making design/engineer collaboration potentially confusing. In addition, this API may be less flexible. What if we want to change the behaviour of one of our tooltip variants? Or introduce a 3rd tooltip usecase.

An API that focuses on what the tooltip usecase is, rather than how it behaves, may be more appropriate. You previously suggested a `type` property, or breaking out into a separate component, could that improve on the points above?

Reviewer: Related, to support the design/engineer workflow ideally naming would be reflected in Figma – same props / component name. Deyan currently documents `o3-toggletip` and `o3-tooltip` but only has 1 Figma component `o3-tooltip` because they look the same. How do we ensure the alignment for smooth handover via dev mode?

https://www.figma.com/file/5ATknbGociZMlnNXX4sy4f/%F0%9F%92%A0-Whitelabel---Design-System?type=design&node-id=4192-618&mode=design&t=5RiBObebw6oEo8Gv-4

If the attribute `render-on-open` is not present the tooltip will be rendered once the target element is hovered, focused or clicked. The tooltip will be removed from the DOM once the target element is no longer hovered, focused or clicked.

Bellow is a markup example of how to use `o3-tooltip` with `render-on-open` attribute:

```html
<div class="o3-brand-whitelabel">
 <!-- DEMO BUTTON ELEMENT TARGETED BY o3-tooltip -->
 <button id="demo-o3-tooltip-id" aria-describedby="o3-tooltip-content">
  Share
 </button>

 <!-- TOOLTIP CUSTOM ELEMENT -->
 <o3-tooltip
  role="tooltip"
  render-on-open
  placement="top"
  target-id="demo-o3-tooltip-id"
  class="o3-tooltip"
 >
  <div class="o3-tooltip-wrapper">
   <div data-tooltip-arrow></div>
   <div class="o3-tooltip-content" id="o3-tooltip-content">

>     <div class="o3-tooltip-content-title">Title</div>

Reviewer: It would be good to explain more what title is and when to use it. I believe there is an accessibility concern here in that we can't emphasis the "title" to screen reader users, so we should be clear that this distinction is presentational only and copy should be written accordingly. Perhaps Deyan can help write those guidelines

    <div>This tooltip renders once DOM is loaded and has close button</div>
   </div>
  </div>
  <button
   class="o3-tooltip-close"
   aria-label="Close tooltip"
   title="Close tooltip"
  ></button>
 </o3-tooltip>
</div>
```

and this is example of how to use `o3-tooltip` without `render-on-open` attribute:

```html
<div class="o3-brand-whitelabel">
 <!-- DEMO BUTTON ELEMENT TARGETED BY o3-tooltip -->
 <button id="demo-o3-tooltip-id" aria-describedby="o3-tooltip-content">
  Share
 </button>

 <!-- TOOLTIP CUSTOM ELEMENT -->
 <o3-tooltip
  role="tooltip"
  placement="top"
  target-id="demo-o3-tooltip-id"
  class="o3-tooltip"
 >
  <div class="o3-tooltip-wrapper">
   <div data-tooltip-arrow></div>
   <div class="o3-tooltip-content" id="o3-tooltip-content">
    <div class="o3-tooltip-content-title">Title</div>
    <div>
     This tooltip renders only when target element is hovered, clicked or
     focused
    </div>
   </div>
  </div>
 </o3-tooltip>
</div>
```

> ### Branding

Reviewer: This is true for every component. We try to avoid duplicating general component use across READMEs. I've onboard with keeping it whilst we have few components though. Where should this content live in the future? A guide on the for everyone site maybe

Author: [ee07873](https://github.com/Financial-Times/origami/pull/1429/commits/ee07873859c69920cd65c9f2811b65a5c825e319) Yeah I agree and we should have a more generic page on o4e page to tell users how to implement branding.

To apply a brand theme wrap your tooltip element inside an element that has a class name of `o3-brand-{yourBrand}`. E.g. to output a tooltip for the professional theme:

```html
<div class="o3-brand-professional">
 <o3-tooltip></o3-tooltip>
</div>
```

You will need to install [o3-web-token](https://www.npmjs.com/package/@financial-times/o3-web-token) package to use the branding feature.

## JavaScript

No code will run automatically unless you import the JavaScript. To import the JavaScript, you can use the following code:

```javascript
import o3Tooltip from '@financial-times/o3-tooltip';
```

## css

To style `o3-tooltip` you can use import brand specific css file from `@financial-times/o3-tooltip` package. You can use the following code:

```javascript

> import '@financial-times/o3-tooltip/css/whitelabel.css';

Reviewer: ```suggestion
import '@financial-times/o3-tooltip/css/[your brand].css';
```
?

Author: [cd7ca27](https://github.com/Financial-Times/origami/pull/1429/commits/cd7ca27b7d9f66f7f9b99c259faea7bce8e69d75) updated all mentions of whitelabel.

```

## JSX

> `o3-tooltip` has a JSX component exported as well. To use it, you can use the following code:

Reviewer: Woop! Glad we're documenting JSX. This is new to most people! Should it come first, above Markup? Or perhaps as part of it, so long as we make the 2 options clear. I think we'd like to encourage more template use.

# o3-tooltip[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](#licence)

> o3-tooltip is hosting two components: `o3-tooltip-onboarding` and `oe-tooltip-toggle`. Onboarding Tooltips help for a smoother introduction to new features or functionality. Toggletips are compact pieces of information designed to offer additional insights when needed.

Reviewer: ```suggestion
o3-tooltip is hosting two components: `o3-tooltip-onboarding` and `o3-tooltip-toggle`. Onboarding Tooltips help for a smoother introduction to new features or functionality. Toggletips are compact pieces of information designed to offer additional insights when needed. See our [design guidelines for usage information and brand support](https://origami-for-everyone.ft.com).
```

(we can update the link when we have it)

- [o3-tooltip](#o3-tooltip)
- [Markup](#markup)
- [JavaScript](#javascript)
- [css](#css)
- [JSX](#jsx)
- [Migration Guide](#migration-guide)
- [Contact](#contact)
- [Licence](#licence)

> ## Markup

Reviewer: I like that you introduce both tooltips initially, with a little overview. For the detail, splitting the 2 types of tooltips into sub-headings may make it easier to find and follow the correct instructions.

> We provide custom elements `<o3-tooltip-onboarding>` and `<o3-tooltip-toggle>` to be used in your markup but we also equivalent provide [JSX components](#jsx) to be used in React.

Reviewer: This suggests we provide a React components – React users will still need to implement custom elements. How about?

```suggestion
We provide custom elements `<o3-tooltip-onboarding>` and `<o3-tooltip-toggle>`. React users may include these using our [JSX template](#jsx).
```

> `<o3-tooltip-onboarding>` is used to display a tooltip on a HTML target element that can have `aria-describedby` attribute. It is used to provide additional information about the target element and is shown immediately once the page is loaded. It can be only closed by clicking the close button.

Reviewer: ```suggestion
`<o3-tooltip-onboarding>` is used to display a tooltip on a HTML target element that can have an `aria-describedby` attribute. It is used to provide additional information about the target element and is shown immediately once the page is loaded. It can be only closed by clicking the close button.
```

> `o3-tooltip-toggle` does not have a close button and is used with info icon. It can be triggered by clicking on the info icon and will be removed from the DOM if user clicks outside of the tooltip, press ESC or click the info icon again.

Reviewer: ```suggestion
`o3-tooltip-toggle` does not have a close button and is used with an information icon. It can be triggered by clicking on the information icon and will be removed from the DOM if the user clicks outside of the tooltip, presses ESC, or clicks the info icon again.
```

> Bellow is a markup examples of how to use `<o3-tooltip-onboarding>` and `<o3-tooltip-toggle>` custom elements:

Reviewer: ```suggestion
Below are examples of how to use `<o3-tooltip-onboarding>` and `<o3-tooltip-toggle>` custom elements:
```

```html
<!-- <o3-tooltip-onboarding> -->
<div data-o3-brand="[your brand]">
 <button
  id="demo-o3-tooltip-id"

>   class="o3-button o3-button--secondary o3-button--big demo-tooltip-target"

Reviewer: Instruction here on what attributes to change would be helpful. In addition, `o3-tooltip-content` could be renamed to more clearly indicate it's for demo purposes

  aria-describedby="o3-tooltip-content"
 >
  Share
 </button>
 <o3-tooltip-onboarding
  role="tooltip"
  placement="top"
  target-id="demo-o3-tooltip-id"
  class="o3-tooltip"
  content="Tool tip content that is quite long, Tool tip content that is quite long, Tool tip content that is quite long"
  title="Title"
  content-id="o3-tooltip-content"
 >
 </o3-tooltip-onboarding>
</div>
```

`content-id` is used to link the tooltip to the target element. It is used to provide additional information about the target element.

```html
<!-- <o3-tooltip-toggle> -->
<div data-o3-brand="[your brand]">
 <o3-tooltip-toggle
  placement="bottom"
  class="o3-tooltip"
  content="click the button to see the tooltip"
  title="Title">
 </o3-tooltip-toggle>
</div>
```

## JavaScript

No code will run automatically unless you import the JavaScript. To import the JavaScript, you can use the following code:

```javascript
import o3Tooltip from '@financial-times/o3-tooltip';
```

## css

> To style `o3-tooltip` you can use import brand specific css file from `@financial-times/o3-tooltip` package. You can use the following code:
> 
> ```javascript
> import '@financial-times/o3-tooltip/css/[your brand].css';
> ```

Reviewer: - This will vary by project. It's good to flag that.
- Most of our users will be using PostCSS / Sass, so a more helpful example might use `@import`?
- There's some duplication here in copy and what you're showing, I believe we can reduce this sentence.

```suggestion
To style `o3-tooltip` import brand specific css, this varies depending on your project. For example:

```scss
@import '@financial-times/o3-tooltip/css/[your brand].css';
```
```

## JSX

> `o3-tooltip` has a two JSX component exported that implement `<o3-tooltip-onboarding>` and `<o3-tooltip-toggle>` under the hood and export `TooltipOnboarding` and `TooltipToggle` respectively. To use it, you can use the following code:

Reviewer: Again this suggests this is an alternative React implementation. I think it's better to avoid `implement` and `component`, in favour of `template` or `elements`:

```suggestion
`o3-tooltip` exports two JSX templates, `TooltipOnboarding` and `TooltipToggle` for each tooltip component. This means copy-pasting html can be avoided. For example:
```

```jsx
import {
 TooltipOnboarding,
 TooltipToggle,

> } from '@financial-times/o3-tooltip/jsx';

Reviewer: As a React user, do I need to take further steps? Perhaps not for this PR, but how do we make that clear?

Author: do you have anything specific in mind here?

Reviewer: Maybe nothing, since all props are strings. Would be a good one to work with Nikolay to try out in one of our production React environments.

- https://react.dev/reference/react-dom/components#custom-html-elements
- https://css-tricks.com/building-interoperable-web-components-react/

import '@financial-times/o3-tooltip'; // import the JavaScript needed for custom elements

import '@financial-times/o3-web-token/[your brand].css'; // brand styling
import '@finacial-times/o3-tooltip/css/[your brand].css'; // tooltip styling

<div className="o3-brand-[your brand]">
 <button id="target" aria-describedby="contentId">
  Target element
 </button>
 <TooltipOnboarding {...props} />;
 <TooltipToggle {...props} />
</div>;
```

### Props

> `TooltipOnboarding` component accept the following props:

Reviewer: ```suggestion
The `TooltipOnboarding` JSX element accepts the following `props`:
```

```ts
type Placement =
 | 'top'
 | 'bottom'
 | 'left'
 | 'right'
 | 'top-start'
 | 'top-end'
 | 'bottom-start'
 | 'bottom-end'
 | 'left-start'
 | 'left-end'
 | 'right-start'
 | 'right-end';
```

|   Prop    |    type     | required | default |             description             |
| :-------: | :---------: | :------: | :-----: | :---------------------------------: |
| targetId  |   string    |   true   |    -    | Target element to attach tooltip to |
|  content  |   string    |   true   |    -    |       Content of the tooltip        |
|   title   |   string    |  false   |    -    |        Title of the tooltip         |
| contentId |   string    |   true   |    -    |  id for `aria-describedby` element  |
| placement | `Placement` |  false   |  'top'  |      Placement of the tooltip       |

> `ToggleTooltip` component accept the following props:

Reviewer: ```suggestion
The `ToggleTooltip` JSX element accepts the following `props`:
```

  aria-describedby="o3-tooltip-content"
 >
  Share
 </button>
 <o3-tooltip-onboarding
  role="tooltip"
  placement="top"
  target-id="demo-o3-tooltip-id"
  class="o3-tooltip"
 >
  <div class="o3-tooltip-wrapper">
   <div data-tooltip-arrow="true"></div>
   <div class="o3-tooltip-content" id="o3-tooltip-content">
    <div class="o3-tooltip-content-title">Title</div>
    <div>
     Tool tip content that is quite long, Tool tip content that is quite
     long, Tool tip content that is quite long
    </div>
   </div>
   <button
    type="button"
    class="o3-tooltip-close"
    aria-label="Close tooltip"
    title="Close tooltip"
   ></button>
  </div>
 </o3-tooltip-onboarding>
</div>
```

```html
<!-- <o3-tooltip-toggle> -->
<div data-o3-brand="[your brand]">
 >
 <o3-tooltip-toggle placement="bottom" class="o3-tooltip">
  <button
   type="button"
   class="o3-tooltip-info"
   aria-label="Tooltip info"
   title="Tooltip description"
  ></button>
  <div class="o3-tooltip-wrapper">
   <div data-tooltip-arrow="true"></div>
   <div class="o3-tooltip-content" role="status">
    <div class="o3-tooltip-content-title">Title</div>
    <div>click the button to see the tooltip</div>
   </div>

>   </div>

Reviewer: We can simplify this interface for users, since we don't need this markup for the core experience.

1. We could create markup in JavaScript
2. Remove excess options, do we need `aria-label` and `title`? 

E.g.
```html
<o3-tooltip-toggle placement="bottom" class="o3-tooltip" targetLabel="Tooltip description">
    click the button to see the tooltip
</o3-tooltip-toggle>
```

Related, what does the core experience look like, when JavaScript isn't available or errors? Is this content hidden?

Reviewer: ☝️ This applies to both types of tooltip

Author: check [891b479](https://github.com/Financial-Times/origami/pull/1429/commits/891b4795bf08be3c9bc44d73065a51316de5af0d)

# o-tooltip [![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](#licence)

o-tooltip is a component usually used for annotating or highlighting bits of user interface.

> - [Usage](#usage)

Author: yeah of course. I am gonna write a new readme for o3-tooltip
